### PR TITLE
BUG: 27453 right merge order

### LINKED ISSUE #27453: Right merge not preserve order
Base by [SO question](https://stackoverflow.com/q/57092373/2901002):

Merge with right join not preserve order of columns:

    columns = ['A', 'B', 'C']
    data_1 = [[2, 5, 3, 5], [8, 2, 4, 1], [6, 5, 9, 1]]
    data_1 = np.array(data_1).T
    df_1 = pd.DataFrame(data=data_1, columns=columns)
    print (df_1)
       A  B  C
    0  2  8  6
    1  5  2  5
    2  3  4  9
    3  5  1  1
    
    columns = ['A', 'B', 'C']
    data_2 = [[2, 5, 3, 5], [7, 1, 3, 0], [np.nan, np.nan, np.nan, np.nan]]
    data_2 = np.array(data_2).T
    df_2 = pd.DataFrame(data=data_2, columns=columns)
    print (df_2)
         A    B   C
    0  2.0  7.0 NaN
    1  5.0  1.0 NaN
    2  3.0  3.0 NaN
    3  5.0  0.0 NaN
    
    #right join NOT preserve order
    df1 = df_1.merge(df_2[['A', 'B']], on=['A', 'B'], how='right')
    print (df1)
       A  B    C
    0  5  1  1.0
    1  2  7  NaN
    2  3  3  NaN
    3  5  0  NaN
    
Correct order for right join:
        
        A   B   C
    0   2   7   NaN
    1   5   1   1
    2   3   3   NaN
    3   5   0   NaN
    
    #left join wpreserve order correct
    df2 = df_1.merge(df_2[['A', 'B']], on=['A', 'B'], how='left')
    print (df2)
       A  B  C
    0  2  8  6
    1  5  2  5
    2  3  4  9
    3  5  1  1

----

    print (pd.show_versions())
    
    INSTALLED VERSIONS
    ------------------
    commit: None
    python: 3.7.3.final.0
    python-bits: 64
    OS: Windows
    OS-release: 7
    machine: AMD64
    processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
    byteorder: little
    LC_ALL: None
    LANG: en
    LOCALE: None.None
    
    pandas: 0.24.2
    pytest: 4.3.1
    pip: 19.0.3
    setuptools: 40.8.0
    Cython: 0.29.6
    numpy: 1.16.2
    scipy: 1.2.1
    pyarrow: None
    xarray: None
    IPython: 7.4.0
    sphinx: 1.8.5
    patsy: 0.5.1
    dateutil: 2.8.0
    pytz: 2018.9
    blosc: None
    bottleneck: 1.2.1
    tables: 3.5.1
    numexpr: 2.6.9
    feather: None
    matplotlib: 3.0.3
    openpyxl: 2.6.1
    xlrd: 1.2.0
    xlwt: 1.3.0
    xlsxwriter: 1.1.5
    lxml.etree: 4.3.2
    bs4: 4.7.1
    html5lib: 1.0.1
    sqlalchemy: 1.3.1
    pymysql: None
    psycopg2: None
    jinja2: 2.10
    s3fs: None
    fastparquet: None
    pandas_gbq: None
    pandas_datareader: None
    gcsfs: None
    None

### PR DESCRIPTION
**Pull request by MarcoGorelli**
Resurrection of #27762

I fixed the NameError in the test, and also changed the example given in the whatsnew entry (I couldn't see a difference between the outputs of 0.25.x and 1.0.0 in the original)

- [x] closes #27453
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by WillAyd** [comment]
lgtm. not sure of the 1.0 milestone at this point but let's see what @TomAugspurger and @jreback think

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -1156,6 +1183,7 @@ Reshaping
 - Bug in :func:`merge_asof` merging on a tz-aware ``left_index`` and ``right_on`` a tz-aware column (:issue:`29864`)
 - Improved error message and docstring in :func:`cut` and :func:`qcut` when `labels=True` (:issue:`13318`)
 - Bug in missing `fill_na` parameter to :meth:`DataFrame.unstack` with list of levels (:issue:`30740`)
+- :meth:`DataFrame.merge` now preserves right frame's row order when executing a right merge (:issue:`27453`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v1.0.0.rst]: Just noticed this had been requested to be removed in the original PR (will wait for a decision at to whether to include this in v1 before putting it through another round of CI)

**Comment by TomAugspurger** [comment]
Is this fixing a regression in 0.25? Otherwise would prefer to keep it for 1.0.1

**Comment by MarcoGorelli** [comment]
@TomAugspurger as in, is it something that was working in 0.25 but isn't working on master? If so, then no, it's not.

Should I make a v1.0.1.rst whatsnew file then?

**Comment by jreback** [comment]
> @TomAugspurger as in, is it something that was working in 0.25 but isn't working on master? If so, then no, it's not a regression in 0.25.
> 
> Should I make a v1.0.1.rst whatsnew file then?

no there is already a PR for that

**Comment by MarcoGorelli** [comment]
> Otherwise would prefer to keep it for 1.0.1

@TomAugspurger sure, whatsnew entry added in v1.0.1

**Comment by WillAyd** [comment]
I think this should be 1.1 at this point? AFAIK not a regression

**Comment by MarcoGorelli** [comment]
> I think this should be 1.1 at this point? AFAIK not a regression

@WillAyd Sure, have moved the whatsnew entry to v1.1.0

**Review by WillAyd — APPROVED** [review]
lgtm

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -1307,15 +1310,20 @@ def _get_join_indexers(
...
     # `count` is the num. of unique keys
     # set(lkey) | set(rkey) == range(count)
-    lkey, rkey, count = _factorize_keys(lkey, rkey, sort=sort)
 
+    # flip left and right keys if performing a right merge
+    # to preserve right merge row order (GH 27453)
+    if how == "right":
+        factorized_rkey, factorized_lkey, count = fkeys(rkey, lkey)
+    else:
+        factorized_lkey, factorized_rkey, count = fkeys(lkey, rkey)
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: add a blank line after

File: pandas/tests/reshape/merge/test_merge.py [hunk 3]
```diff
@@ -2167,3 +2167,35 @@ def test_merge_datetime_upcast_dtype():
         }
     )
     tm.assert_frame_equal(result, expected)
+
+
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py]: can you *also* add the example you have in the whatsnew

File: pandas/tests/reshape/merge/test_merge.py [hunk 4]
```diff
@@ -2167,3 +2167,35 @@ def test_merge_datetime_upcast_dtype():
...
+@pytest.mark.parametrize("how", ["left", "right"])
+def test_merge_preserves_row_order(how):
+    # GH 27453
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py]: is this the example from the OP (I would rather use that one exactly)

File: pandas/tests/reshape/merge/test_merge.py [hunk 5]
```diff
@@ -2167,3 +2167,35 @@ def test_merge_datetime_upcast_dtype():
...
+    population_df = DataFrame(
+        {
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py]: cam you also add an example where merging on the index (both left and right) with how='right' (and 'left') for that matter

File: pandas/tests/reshape/merge/test_merge.py [hunk 6]
```diff
@@ -2167,3 +2167,35 @@ def test_merge_datetime_upcast_dtype():
         }
     )
     tm.assert_frame_equal(result, expected)
+
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py]: can you locate this near any other tests for right merge (or may not be any, if so pls comment)

File: pandas/core/reshape/merge.py [hunk 7]
```diff
@@ -1307,15 +1310,20 @@ def _get_join_indexers(
     # factorize keys to a dense i8 space
     # `count` is the num. of unique keys
     # set(lkey) | set(rkey) == range(count)
-    lkey, rkey, count = _factorize_keys(lkey, rkey, sort=sort)
 
+    # flip left and right keys if performing a right merge
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: so another possiblity is to pass how= to _factorize_keys, this *maybe* a more complicated code path; pls comment one way or the other (in the PR).

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/merge.py (hunk 7 above)
Comment (MarcoGorelli) [reply pandas/core/reshape/merge.py]: @jreback Sure, I've gone with that and it seems a lot simpler

**Comment by MarcoGorelli** [comment]
ping :)

File: pandas/core/reshape/merge.py [hunk 8]
```diff
@@ -1847,7 +1847,7 @@ def _right_outer_join(x, y, max_groups):
     return left_indexer, right_indexer
 
 
-def _factorize_keys(lk, rk, sort=True):
+def _factorize_keys(lk, rk, sort=True, how="inner"):
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: can you add a doc-string and typing here

File: pandas/tests/reshape/merge/test_merge.py [hunk 9]
```diff
@@ -1318,6 +1318,44 @@ def test_merge_right_index_right(self):
...
+        result = df1.merge(df2[["A", "B"]], on=["A", "B"], how=how)
+        expected = pd.DataFrame({"A": a})
+        if how == "right":
+            expected["B"] = df2["B"]
+        else:
+            expected["B"] = df1["B"]
+        tm.assert_frame_equal(result, expected)
+
+        left_df = pd.DataFrame({"colors": ["blue", "red"]}, index=pd.Index([0, 1]))
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py]: either reduce the examples here (they look like 3x of the same one), or parameterize over them.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.1.0.rst [hunk 10]
```diff
@@ -168,6 +168,32 @@ key and type of :class:`Index`.  These now consistently raise ``KeyError`` (:iss
...
+*pandas 1.0.x*
+
+.. code-block:: python
+
+    >>> left_df.merge(right_df, on=['animal', 'max_speed'], how="right")
+        animal  max_speed
+    0      pig         11
+    1  quetzal         80
+
+*pandas 1.1.0*
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: current behavior

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. some very minor doc comments, ping on green.

**Comment by MarcoGorelli** [comment]
@jreback thanks - done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @MarcoGorelli for taking this over the line!

**Pull request merged**